feat: New Outbox Contract #4768

sklppy88 opened the pull request:
Resolves #4768

Commit:
initial

Commit:
new

Commit:
fix name

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
fix

Commit:
Add error cases

Commit:
fix

Commit:
fomrat

fix

comments

Commit:
slither

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
slither

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
correct slither version

nventuro commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,225 @@
+    uint256 _blockNumber
+  ) public {
+    // We most likely will not have > 128 L2 -> L1 Messages in a single block but are testing an upper bound here
+    vm.assume(_recipients.length < 256);
+    vm.assume(_recipients.length > 128);
```
See my prior comment about `vm.assume`, particularly if you're having performance issues.

nventuro commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,225 @@
+    vm.expectRevert(abi.encodeWithSelector(Errors.Outbox__RootAlreadySet.selector, 1));
+    outbox.insert(1, root, _DEFAULT_TREE_HEIGHT);
+  }
+
```
Same with `consume` - we only have a single test case (the multiple recipients one) and it's quite convoluted. Consider adding a simple one where you test events, the nullified getter, etc.

nventuro reviewed (commented):
Went a bit overboard with the comments :sweat_smile: Hope you find them useful. Overall it looks great! Most of my suggestions are quite minor.

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
add fn selectors

comments

comments

fix

fix

comments

comments

comments

test

test

comment

https://github.com/ethereum/solidity/issues/14430

comments

test

comments

liberryfy

test

asdf

add manual test

asdf

test

fix

asdf

asdf

comments

comments

LHerskind commented on l1-contracts/src/core/interfaces/messagebridge/INewOutbox.sol:
```diff
@@ -0,0 +1,46 @@
+   * @param _l2BlockNumber - The block number specifying the block that contains the message we want to consume
+   * @param _leafIndex - The index inside the merkle tree where the message is located
+   * @param _message - The L2 to L1 message
+   * @param _path - The sibling path used to prove inclusion of the message
```
The length of the path "depends", but could mark that it depends on the block size and from leaf to root 👍

LHerskind commented on l1-contracts/src/core/libraries/Errors.sol:
```diff
@@ -46,6 +46,12 @@ library Errors {
     uint32 storedDeadline,
     uint32 deadlinePassed
   ); // 0x5e789f34
+  error Outbox__InvalidPathLength(uint256 expected, uint256 actual);
+  error Outbox__InvalidRoot(bytes32 expected, bytes32 actual);
+  error Outbox__RootAlreadySet(uint256 l2BlockNumber);
+  error Outbox__InvalidRecipient(address expected, address actual);
+  error Outbox__AlreadyNullified(uint256 l2BlockNumber, uint256 leafIndex);
+  error Outbox__NothingToConsumeAtBlock(uint256 l2BlockNumber);
```
The selectors needs to be added yes. You can use cast for it. Example:
```bash
cast sig "Outbox__NothingToConsumeAtBlock(uint256 l2BlockNumber)"
```

LHerskind commented on l1-contracts/src/core/interfaces/messagebridge/INewOutbox.sol:
```diff
@@ -0,0 +1,46 @@
+
+  /*
+   * @notice Inserts the root of a merkle tree containing all of the L2 to L1 messages in
+   * a block specified by _blockNumber.
```
The comment uses `_blockNumber` but not such parameter is part of the function.

LHerskind commented on l1-contracts/src/core/messagebridge/NewOutbox.sol:
```diff
@@ -0,0 +1,144 @@
+      revert Errors.Outbox__Unauthorized();
+    }
+
+    if (roots[_l2BlockNumber].root != bytes32(0)) {
+      revert Errors.Outbox__RootAlreadySet(_l2BlockNumber);
+    }
```
The rollup is the only that can insert and would always be inserting non-zero, but seems good to just make it explicit here. Add it as the first check before any state is read just to short-circuit in case.

LHerskind commented on l1-contracts/test/merkle/Merkle.t.sol:
```diff
@@ -27,4 +27,28 @@ contract MerkleTest is Test {
+      emit log_named_bytes32("path", path[i]);
+    }
+
+    emit log_named_bytes32("test", leaf);
+    emit log_named_bytes32("size", bytes32(merkle.SIZE()));
+    emit log_named_bytes32("root", merkle.computeRoot());
```
I think the logs are just leftovers? Would remove them entirely form the test.

LHerskind commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,225 @@
+    outbox.consume(1, 0, fakeMessage, path);
+  }
+
+  function testInsertVariedLeafs(bytes32[] calldata _messageLeafs) public {
+    vm.assume(_messageLeafs.length < 256);
+    vm.assume(_messageLeafs.length > 128);
```
The bound is a function on the test, so you just call `bound`, no `vm` needed.

LHerskind reviewed (commented):
@nventuro had some good points around extra tests and clarity. Added a few notes.

Commit:
Merge remote-tracking branch 'origin/master' into ek/feat/new-outbox-contract

Commit:
slither

Commit:
fix

Commit:
use 8.22

Commit:
comment

Commit:
comments

Commit:
remove

Commit:
format

Commit:
lint

Commit:
slither

Commit:
upgrade to .24

Commit:
fix

Commit:
slither

Commit:
test

nventuro commented on l1-contracts/src/core/libraries/Merkle.sol:
```diff
@@ -0,0 +1,73 @@
+  }
+
+  /*
+  * @notice Calculates a tree height from the amount of elements in the tree, essentially 1 if x = 0 | 1, otherwise Math.ceil(Math.log2(x))
```
I'd leave the impl details out of the description, and instead expand more on the implementation with in-line comments.

nventuro commented on l1-contracts/src/core/libraries/Merkle.sol:
```diff
@@ -0,0 +1,73 @@
+      return 1;
+    }
+
+    uint256 originalNumber = _size;
+
+    uint256 height = 0;
+    while (_size > 1) {
+      _size >>= 1;
+      height++;
+    }
+
+    return 2 ** height != originalNumber ? ++height : height;
```
I found this quite hard to follow, specially since a) I'm not sure what you're trying to do here, b) `originalNumber` is the original size, the whole rename is weird, and I don't follow why you compare size and height, c) the final expression is fairly involed.

nventuro commented on l1-contracts/src/core/messagebridge/NewOutbox.sol:
```diff
@@ -0,0 +1,116 @@
+
+    RootData storage rootData = roots[_l2BlockNumber];
+
+    bytes32 expectedRoot = rootData.root;
```
I'd rename this to e.g. block root. The fact that it's 'expected' is more of a verfication concern, I found it odd to think about verification so early in the lifecycle here.

nventuro replied:
`blockRoot` and `treeHeight` are fine I think. Height is used in a single place, so it's meaning is quite clear.

nventuro commented on l1-contracts/src/core/messagebridge/NewOutbox.sol:
```diff
@@ -0,0 +1,116 @@
+      revert Errors.Outbox__AlreadyNullified(_l2BlockNumber, _leafIndex);
+    }
+
+    uint256 expectedHeight = rootData.height;
+
+    if (expectedHeight != _path.length) {
+      revert Errors.Outbox__InvalidPathLength(expectedHeight, _path.length);
+    }
```
Similarly this height is not 'expected' really.

nventuro commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,237 @@
+
+    vm.expectEmit(true, true, true, true, address(outbox));
+    emit INewOutbox.MessageConsumed(1, root, leaf, 0);
+    outbox.consume(1, 0, fakeMessage, path);
```
I'd also check here that the nullifier is set.

nventuro commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,237 @@
+    outbox.consume(1, 0, fakeMessage, path);
+  }
+
+  // This test takes awhile so to keep it somewhat reasonable we've set a limit on the amount of fuzz runs
+  /// forge-config: default.fuzz.runs = 32
```
Is this still true after the removal of the `vm.assume` calls?

nventuro replied:
Can you make the `recipients` array fixed-size, and then pick a a random length using a `size` variable that we bound? Otherwise we don't really know why many recipients we're testing for.

nventuro commented on l1-contracts/test/merkle/Merkle.t.sol:
```diff
@@ -27,4 +51,96 @@ contract MerkleTest is Test {
+    assertEq(path2[2], expectedPath2[2]);
+  }
+
+  function testCalculateNextHighestPowerOfTwo() external {
```
```suggestion
  function testCalculateTreeHeightFromSize() external {
```

Commit:
new detail

sklppy88 committed:
Update l1-contracts/test/merkle/Merkle.t.sol

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>

sklppy88 committed:
Update l1-contracts/test/merkle/Merkle.t.sol

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>

Commit:
naming

Commit:
fix

Commit:
comments

Commit:
test

Commit:
comments

Commit:
asdf

Commit:
fix

test

asdf

asdf

format

nventuro commented on l1-contracts/src/core/libraries/Merkle.sol:
```diff
@@ -0,0 +1,53 @@
+    * @remark - We use the indexAtHeight to see whether our child of the next subtree is at the left or the right side, as any odd indexes are right-sided children.
+    * This is important and we have assigned it into its own variable isRight, because this affects the way we concat our two children to then hash and calculate the root.
+    */
+  function verifyMembership(
+    bytes32[] memory _path,
+    bytes32 _leaf,
+    uint256 _index,
+    bytes32 _expectedRoot
+  ) internal pure {
```
Well an invalid index would be a different kind of error that'd point at either the index being wrong or the path being constructed incorrectly. But I agree, we don't need this.

Commit:
fix

Commit:
comments

Commit:
slith

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
test

benesjan commented on l1-contracts/src/core/messagebridge/NewOutbox.sol:
```diff
@@ -0,0 +1,131 @@
+/**
+ * @title NewOutbox
+ * @author Aztec Labs
+ * @notice Lives on L1 and is used to consume L2 -> L1 messages. Messages are inserted by the state transitioner
```
Everywhere else in solidity we call state transitioner a Rollup so would prefer to stick to that naming. I see Rollup as an implementation of the abstract "state transitioner" from yellow paper.

benesjan commented on l1-contracts/src/core/messagebridge/NewOutbox.sol:
```diff
@@ -0,0 +1,131 @@
+    uint256 _l2BlockNumber,
+    uint256 _leafIndex,
+    DataStructures.L2ToL1Msg memory _message,
+    bytes32[] memory _path
```
It seems to me that it would be better to use calldata for _message and _path given that you don't need to modify the values and calldata is cheaper.

nventuro replied:
(do benchmark and compare if you do this though)

LHerskind replied:
It is not a speedup that you can just see, it is evaluating the gas, you can add `--gas-report` to see the gas, and compare those. Think there is a new flag now to do it properly though for the exact gas but cannot remember it at the top of my head.

benesjan commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,250 @@
+  address internal constant _NOT_STATE_TRANSITIONER = address(0x69);
+  address internal constant _NOT_RECIPIENT = address(0x420);
+  uint256 internal constant _DEFAULT_TREE_HEIGHT = 2;
+  uint256 internal constant _VERSION = 0;
```
We don't currently use the underscore for constants in our codebase so would keep that consistent.

benesjan commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,250 @@
+    outbox.insert(1, root, _DEFAULT_TREE_HEIGHT);
+  }
+
+  function testInsertVariedLeafs(bytes32[] calldata _messageLeafs) public {
```
Could you add a comment here describing what this is testing? It's not clear to me form the name.

benesjan commented on l1-contracts/test/merkle/Merkle.t.sol:
```diff
@@ -5,17 +5,41 @@ pragma solidity >=0.8.18;
 import {FrontierMerkle} from "./../../src/core/messagebridge/frontier_tree/Frontier.sol";
 
-contract MerkleTest is Test {
+abstract contract MerkleLibWrapper {
```
I think readability would be better here if you used the helper approach like what we used for example in TxsDecoderHelper. Having this abstract contract and then inheriting from it in the test seems less readable to me than having test code and the library separated in different contracts.

benesjan commented on l1-contracts/test/merkle/Merkle.t.sol:
```diff
@@ -5,17 +5,41 @@ pragma solidity >=0.8.18;
+    merkle = new NaiveMerkle(DEPTH);
+
+    // Set up testNaive
+    testNaiveMerkle = new NaiveMerkle(DEPTH);
```
Having merkle and testNaiveMerkle next too each other here is confusing. I see that merkle is only used in frontier tree test so I would separate the tests here to Frontier.t.sol and Merkle.t.sol and would mention in Merkle.t.sol that it's testing only the Merkle library.

benesjan commented on l1-contracts/test/merkle/Merkle.t.sol:
```diff
@@ -27,4 +51,125 @@ contract MerkleTest is Test {
     }
   }
+
+  function testNaiveSiblingPathAndMembershipVerification(uint256 _idx) public view {
```
The focus here is on testing that the membership verification works and not on the naive tree so would just call this testVerifyMembership.

benesjan commented on l1-contracts/test/merkle/Merkle.t.sol:
```diff
@@ -27,4 +51,125 @@ contract MerkleTest is Test {
+    vm.expectRevert();
+    this.verifyMembershipWrapper(truncatedPath, leaf, leafIndex, expectedRoot);
+
+    // Tests empty path
```
I don't think having these 3 cases of empty, truncated and garbled path here really makes sense given the nature of merkle trees. Instead I would just use fuzzing and make the fuzzer provide the path on the input.

nventuro replied:
All the fuzzer tests generate passing cases though, because we manually compute the correct path. I think it's important to have _some_ failure cases.

benesjan reviewed (changes requested):
Mostly just minor things

sklppy88 committed:
Update l1-contracts/src/core/libraries/Merkle.sol

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>

Commit:
move out of core

add comments

comments

comments

fix

fix

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
abc

test

fix

move

move

fix

test

Commit:
format

Commit:
slither

Commit:
naming

Commit:
renames

Commit:
slith

LHerskind commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,254 @@
+  address internal constant NOT_ROLLUP_CONTRACT = address(0x69);
+  address internal constant NOT_RECIPIENT = address(0x420);
+  uint256 internal constant DEFAULT_TREE_HEIGHT = 2;
+  uint256 internal constant AZTEC_VERSION = 0;
```
As I recall the registry is using 0 for the "nothing" empty case and then only inserting the next thing as version 1. Reasoning was that having zero for the value could lead to pain inside the circuit (back in the days) so we started it from 1.

Commit:
Version update

Commit:
rename

Commit:
comments

Commit:
comments

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

spalladino committed:
fix: Point docs links to current tag if available (#5219)

Instead of pointing to master, load the current tag if it's available on
the environment and use it for redirecting to github.

critesjosh committed:
chore(docs): Add details to getting started contract deployment (#5220)

Add some feedback based on office hours today.

AztecBot committed:
chore(master): Release 0.28.0 (#5192)

:robot: I have created a release *beep* *boop*
---


<details><summary>aztec-package: 0.28.0</summary>

##
[0.28.0](https://github.com/AztecProtocol/aztec-packages/compare/aztec-package-v0.27.2...aztec-package-v0.28.0)
(2024-03-14)


### ⚠ BREAKING CHANGES

* Support contracts with no constructor
([#5175](https://github.com/AztecProtocol/aztec-packages/issues/5175))

### Features

* Support contracts with no constructor
([#5175](https://github.com/AztecProtocol/aztec-packages/issues/5175))
([df7fa32](https://github.com/AztecProtocol/aztec-packages/commit/df7fa32f34e790231e091c38a4a6e84be5407763))
</details>

<details><summary>barretenberg.js: 0.28.0</summary>

##
[0.28.0](https://github.com/AztecProtocol/aztec-packages/compare/barretenberg.js-v0.27.2...barretenberg.js-v0.28.0)
(2024-03-14)


### Miscellaneous

* **barretenberg.js:** Synchronize aztec-packages versions
</details>

<details><summary>aztec-cli: 0.28.0</summary>

##
[0.28.0](https://github.com/AztecProtocol/aztec-packages/compare/aztec-cli-v0.27.2...aztec-cli-v0.28.0)
(2024-03-14)


### ⚠ BREAKING CHANGES

* Support contracts with no constructor
([#5175](https://github.com/AztecProtocol/aztec-packages/issues/5175))

### Features

* Support contracts with no constructor
([#5175](https://github.com/AztecProtocol/aztec-packages/issues/5175))
([df7fa32](https://github.com/AztecProtocol/aztec-packages/commit/df7fa32f34e790231e091c38a4a6e84be5407763))
</details>

<details><summary>aztec-packages: 0.28.0</summary>

##
[0.28.0](https://github.com/AztecProtocol/aztec-packages/compare/aztec-packages-v0.27.2...aztec-packages-v0.28.0)
(2024-03-14)


### ⚠ BREAKING CHANGES

* Support contracts with no constructor
([#5175](https://github.com/AztecProtocol/aztec-packages/issues/5175))

### Features

* **avm-simulator:** Euclidean and field div
([#5181](https://github.com/AztecProtocol/aztec-packages/issues/5181))
([037a38f](https://github.com/AztecProtocol/aztec-packages/commit/037a38f498ee7f9d9c530a4b3b236e9c377b377d))
* Isolate Plonk dependencies
([#5068](https://github.com/AztecProtocol/aztec-packages/issues/5068))
([5cbbd7d](https://github.com/AztecProtocol/aztec-packages/commit/5cbbd7da89488f6f662f96d0a3532921534755b4))
* New brillig field operations and refactor of binary operations
([#5208](https://github.com/AztecProtocol/aztec-packages/issues/5208))
([eb69504](https://github.com/AztecProtocol/aztec-packages/commit/eb6950462b1ab2a0c8f50722791c7b0b9f1daf83))
* Parallelize linearly dependent contribution in PG
([#4742](https://github.com/AztecProtocol/aztec-packages/issues/4742))
([d1799ae](https://github.com/AztecProtocol/aztec-packages/commit/d1799aeccb328582fabed25811e756bf0453216c))
* Parity circuits
([#5082](https://github.com/AztecProtocol/aztec-packages/issues/5082))
([335c46e](https://github.com/AztecProtocol/aztec-packages/commit/335c46e7b7eddc0396190e6dae7eb2255e3caa9e))
* Support contracts with no constructor
([#5175](https://github.com/AztecProtocol/aztec-packages/issues/5175))
([df7fa32](https://github.com/AztecProtocol/aztec-packages/commit/df7fa32f34e790231e091c38a4a6e84be5407763))
* Track side effects in public
([#5129](https://github.com/AztecProtocol/aztec-packages/issues/5129))
([d666f6f](https://github.com/AztecProtocol/aztec-packages/commit/d666f6f1a0a67fd95694bb5f42b3e7af19a0abea)),
closes
[#5185](https://github.com/AztecProtocol/aztec-packages/issues/5185)
* Update SMT Circuit class and add gate relaxation functionality
([#5176](https://github.com/AztecProtocol/aztec-packages/issues/5176))
([5948996](https://github.com/AztecProtocol/aztec-packages/commit/5948996c0bab8ee99c4686352b8475da38604f28))


### Bug Fixes

* **avm-transpiler:** FDIV and U128 test case
([#5200](https://github.com/AztecProtocol/aztec-packages/issues/5200))
([6977e81](https://github.com/AztecProtocol/aztec-packages/commit/6977e8166b5c27685458a6e04e840b45a77d4765))
* Barretenberg-acir-tests-bb.js thru version bump
([#5216](https://github.com/AztecProtocol/aztec-packages/issues/5216))
([9298f93](https://github.com/AztecProtocol/aztec-packages/commit/9298f932b2d22aa5a4c87dab90d5e72614f222da))
* Do not release docs on every commit to master
([#5214](https://github.com/AztecProtocol/aztec-packages/issues/5214))
([c34a299](https://github.com/AztecProtocol/aztec-packages/commit/c34a299e354847e3e4e253b41921814e86b38645))
* Fail transaction if we revert in setup or teardown
([#5093](https://github.com/AztecProtocol/aztec-packages/issues/5093))
([db9a960](https://github.com/AztecProtocol/aztec-packages/commit/db9a960a99db663a328b261e08917ce5f1dd4e69))
* Intermittent invert 0 in Goblin
([#5189](https://github.com/AztecProtocol/aztec-packages/issues/5189))
([6c70624](https://github.com/AztecProtocol/aztec-packages/commit/6c7062443ae23cc75ac06b7ac1492d12f803d0e5))
* Point docs links to current tag if available
([#5219](https://github.com/AztecProtocol/aztec-packages/issues/5219))
([0e9c7c7](https://github.com/AztecProtocol/aztec-packages/commit/0e9c7c757ed5501d01bb20a57f22e857cf50b93d))
* Remove embedded srs
([#5173](https://github.com/AztecProtocol/aztec-packages/issues/5173))
([cfd673d](https://github.com/AztecProtocol/aztec-packages/commit/cfd673d6224e95a7b09eaa51e1f6535b277b2827))
* Split setup/teardown functions when there's no public app logic
([#5156](https://github.com/AztecProtocol/aztec-packages/issues/5156))
([2ee13b3](https://github.com/AztecProtocol/aztec-packages/commit/2ee13b3e9d17d4715ec72c738cf74e75e3c1581f))
* Validate EthAddress size in aztec-nr
([#5198](https://github.com/AztecProtocol/aztec-packages/issues/5198))
([201c5e1](https://github.com/AztecProtocol/aztec-packages/commit/201c5e1cf94b448f4f75f460a9838b526903f3ce))


### Miscellaneous

* Add dependency instructions to bberg README
([#5187](https://github.com/AztecProtocol/aztec-packages/issues/5187))
([850febc](https://github.com/AztecProtocol/aztec-packages/commit/850febc31400b0f5ca2064d91833a847adc5df31))
* **avm-simulator:** Make sure we support Map storage
([#5207](https://github.com/AztecProtocol/aztec-packages/issues/5207))
([08835f9](https://github.com/AztecProtocol/aztec-packages/commit/08835f99e11c479cb498b411b15a16305695039f))
* **avm-simulator:** Restructure contract storage tests
([#5194](https://github.com/AztecProtocol/aztec-packages/issues/5194))
([fcdd1cc](https://github.com/AztecProtocol/aztec-packages/commit/fcdd1cc260c1faf14eb5fe719d5c7f5306699b1e))
* **docs:** Add details to getting started contract deployment
([#5220](https://github.com/AztecProtocol/aztec-packages/issues/5220))
([5c267ae](https://github.com/AztecProtocol/aztec-packages/commit/5c267ae50561c36eb02b84e5f8f7043b929e906c))
* Moving wit comms and witness and comm labels from instance to oink
([#5199](https://github.com/AztecProtocol/aztec-packages/issues/5199))
([19eb7f9](https://github.com/AztecProtocol/aztec-packages/commit/19eb7f9bd48f1f5fb8d9e9a2e172c8f0c2c9445b))
* Oink
([#5210](https://github.com/AztecProtocol/aztec-packages/issues/5210))
([321f149](https://github.com/AztecProtocol/aztec-packages/commit/321f149dd720f2e74d3b4118bf75c910b466d0ed))
* Pull noir
([#5193](https://github.com/AztecProtocol/aztec-packages/issues/5193))
([aa90f6e](https://github.com/AztecProtocol/aztec-packages/commit/aa90f6ed7bfae06bdf6990816d154bbd24993689))
* Trying to fix intermitent ci failure for boxes
([#5182](https://github.com/AztecProtocol/aztec-packages/issues/5182))
([f988cb8](https://github.com/AztecProtocol/aztec-packages/commit/f988cb85a35fbc16690c81071d8153bd76c51185))


### Documentation

* **yellow-paper:** Add pseudocode for verifying broadcasted functions
in contract deployment
([#4431](https://github.com/AztecProtocol/aztec-packages/issues/4431))
([8bdb921](https://github.com/AztecProtocol/aztec-packages/commit/8bdb9213ff2560a83aadd7cc4af062e08e98bd22))
</details>

<details><summary>barretenberg: 0.28.0</summary>

##
[0.28.0](https://github.com/AztecProtocol/aztec-packages/compare/barretenberg-v0.27.2...barretenberg-v0.28.0)
(2024-03-14)


### Features

* **avm-simulator:** Euclidean and field div
([#5181](https://github.com/AztecProtocol/aztec-packages/issues/5181))
([037a38f](https://github.com/AztecProtocol/aztec-packages/commit/037a38f498ee7f9d9c530a4b3b236e9c377b377d))
* Isolate Plonk dependencies
([#5068](https://github.com/AztecProtocol/aztec-packages/issues/5068))
([5cbbd7d](https://github.com/AztecProtocol/aztec-packages/commit/5cbbd7da89488f6f662f96d0a3532921534755b4))
* New brillig field operations and refactor of binary operations
([#5208](https://github.com/AztecProtocol/aztec-packages/issues/5208))
([eb69504](https://github.com/AztecProtocol/aztec-packages/commit/eb6950462b1ab2a0c8f50722791c7b0b9f1daf83))
* Parallelize linearly dependent contribution in PG
([#4742](https://github.com/AztecProtocol/aztec-packages/issues/4742))
([d1799ae](https://github.com/AztecProtocol/aztec-packages/commit/d1799aeccb328582fabed25811e756bf0453216c))
* Update SMT Circuit class and add gate relaxation functionality
([#5176](https://github.com/AztecProtocol/aztec-packages/issues/5176))
([5948996](https://github.com/AztecProtocol/aztec-packages/commit/5948996c0bab8ee99c4686352b8475da38604f28))


### Bug Fixes

* Barretenberg-acir-tests-bb.js thru version bump
([#5216](https://github.com/AztecProtocol/aztec-packages/issues/5216))
([9298f93](https://github.com/AztecProtocol/aztec-packages/commit/9298f932b2d22aa5a4c87dab90d5e72614f222da))
* Intermittent invert 0 in Goblin
([#5189](https://github.com/AztecProtocol/aztec-packages/issues/5189))
([6c70624](https://github.com/AztecProtocol/aztec-packages/commit/6c7062443ae23cc75ac06b7ac1492d12f803d0e5))
* Remove embedded srs
([#5173](https://github.com/AztecProtocol/aztec-packages/issues/5173))
([cfd673d](https://github.com/AztecProtocol/aztec-packages/commit/cfd673d6224e95a7b09eaa51e1f6535b277b2827))


### Miscellaneous

* Add dependency instructions to bberg README
([#5187](https://github.com/AztecProtocol/aztec-packages/issues/5187))
([850febc](https://github.com/AztecProtocol/aztec-packages/commit/850febc31400b0f5ca2064d91833a847adc5df31))
* Moving wit comms and witness and comm labels from instance to oink
([#5199](https://github.com/AztecProtocol/aztec-packages/issues/5199))
([19eb7f9](https://github.com/AztecProtocol/aztec-packages/commit/19eb7f9bd48f1f5fb8d9e9a2e172c8f0c2c9445b))
* Oink
([#5210](https://github.com/AztecProtocol/aztec-packages/issues/5210))
([321f149](https://github.com/AztecProtocol/aztec-packages/commit/321f149dd720f2e74d3b4118bf75c910b466d0ed))
</details>

---
This PR was generated with [Release
Please](https://github.com/googleapis/release-please). See
[documentation](https://github.com/googleapis/release-please#release-please).

spalladino committed:
fix: Docs links URL missing a letter (#5223)

Seems like COMMIT_TAG strips the leading `v` as well. This seems to
contradict what we're seeing here, but oh well.


https://github.com/AztecProtocol/aztec-packages/blob/37bdc1820f92ad2cc545e6843a0dd90273c21e0e/build-system/scripts/setup_env#L17

critesjosh committed:
chore(docs): add note on new initializer (#5224)

updates the getting started guide to discuss initializer instead of
constructor

critesjosh committed:
fix(docs): Update other constructor refs in docs to use initializer (#5227)

Please read [contributing guidelines](CONTRIBUTING.md) and remove this
line.

---------

Co-authored-by: Santiago Palladino <santiago@aztecprotocol.com>

sklppy88 committed:
Update INewOutbox.sol

benesjan commented on l1-contracts/test/NewOutbox.t.sol:
```diff
@@ -0,0 +1,254 @@
+  function setUp() public {
+    outbox = new NewOutbox(ROLLUP_CONTRACT);
+    zeroedTree = new NaiveMerkle(DEFAULT_TREE_HEIGHT);
+    merkleLibTest = new MerkleLibTest();
```
I think it's a bit weird to instantiate the whole foundry test file like this just to get access to `calculateTreeHeightFromSize` function so I would prefer to just create a separate file with the util function and import just that but it's not a big deal.

benesjan reviewed (approved):
Looks good. Just minor nits

sklppy88 committed:
Update l1-contracts/src/core/messagebridge/NewOutbox.sol

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>

sklppy88 committed:
Update l1-contracts/src/core/messagebridge/NewOutbox.sol

Co-authored-by: Jan Beneš <janbenes1234@gmail.com>

Commit:
move to util

Commit:
test

Commit:
slither

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract

Commit:
Merge branch 'master' into ek/feat/new-outbox-contract
